Source/SwiftLintFramework/Rules/Style/SwitchCaseSortRule.swift

```diff
@@ -0,0 +1,243 @@
+            }
+
+            let sortedCaseNames = caseNamesToSort.sorted()
+            if caseNamesToSort != sortedCaseNames {
+                violations.append(node.switchKeyword.positionAfterSkippingLeadingTrivia)
+            }
+        }
+    }
+}
+
+private extension SwitchCaseSortRule {
+    final class Rewriter: SyntaxRewriter, ViolationsSyntaxRewriter {
+        private(set) var correctionPositions: [AbsolutePosition] = []
+        private let locationConverter: SourceLocationConverter
+        private let disabledRegions: [SourceRange]
+
+        init(locationConverter: SourceLocationConverter, disabledRegions: [SourceRange]) {
+            self.locationConverter = locationConverter
+            self.disabledRegions = disabledRegions
+        }
+
+        override func visit(_ node: SwitchExprSyntax) -> ExprSyntax {
+            guard node.cases.isNotEmpty,
+                  !node.isContainedIn(regions: disabledRegions, locationConverter: locationConverter)
+            else {
+                return super.visit(node)
+            }
+
+            let newCases = sort(node.cases)
+            if node.cases.description != newCases.description {
+                var newNode = node
+                newNode.cases = newCases
+                correctionPositions.append(node.switchKeyword.positionAfterSkippingLeadingTrivia)
+                return super.visit(newNode)
+            }
+
+            return super.visit(node)
+        }
+    }
+}
+
+private func sort(_ cases: SwitchCaseListSyntax) -> SwitchCaseListSyntax {
+    var elements = [SwitchCaseListSyntax.Element]()
+    for caseListElement in cases {
+        // only sort normal switch cases, not the ones with #if...#endif
+        guard case let .switchCase(caseSyntax) = caseListElement else {
+            return cases
+        }
+        // if it's a case statement with multiple items sort them inside first
+        if case let .case(caseLabelSyntax) = caseSyntax.label,
+           !caseLabelSyntax.caseItems.isEmpty {
+            elements.append(.switchCase(caseSyntax.with(\.label, .case(sort(caseLabelSyntax)))))
+        } else { // default or single cases
+            elements.append(caseListElement)
+        }
+    }
+
+    return SwitchCaseListSyntax(elements.sorted(by: byName))
+}
+
+private func sort(_ caseLabel: SwitchCaseLabelSyntax) -> SwitchCaseLabelSyntax {
+    // remove trailing commas, then sort them, then join them with commas
+    let sortedItems = caseLabel.caseItems
+        .map { $0.with(\.trailingComma, nil) }
+        .sorted(by: byName)
+        .enumerated()
+        .map { index, item in
+            index == caseLabel.caseItems.count - 1 ? item : item.with(\.trailingComma, ", ")
+        }
+    return caseLabel.with(\.caseItems, CaseItemListSyntax(sortedItems))
+}
+
+private func sortableName(for expressionPattern: ExpressionPatternSyntax) -> String? {
+    if let memberAccessExpression = expressionPattern.expression
```

Comment:
Could be implemented as a `switch` statement with type-casting patterns. Or see my other comment mentioning `SyntaxTransformVisitor`.